plugins/nordic-hid/fu-nordic-hid-cfg-channel.c

```diff
@@ -0,0 +1,1054 @@
+
+	chunk_len = fu_chunk_get_data_sz(chk);
+	while (offset < chunk_len) {
+		guint8 data_len;
+		guint8 data[REPORT_DATA_MAX_LEN] = {0};
+		g_autoptr(FuNordicCfgChannelMsg) res = g_new0(FuNordicCfgChannelMsg, 1);
+
+		data_len = ((offset + REPORT_DATA_MAX_LEN) < chunk_len)
+			       ? REPORT_DATA_MAX_LEN
+			       : (guint8)(chunk_len - offset);
+
+		if (!fu_memcpy_safe(data,
+				    REPORT_DATA_MAX_LEN,
+				    0,
+				    fu_chunk_get_data(chk),
+				    chunk_len,
+				    offset,
+				    data_len,
+				    error)) {
+			return FALSE;
+		}
+
+		if (!fu_nordic_hid_cfg_channel_cmd_send(self,
+							self->peer_id,
+							"dfu",
+							"data",
+							CONFIG_STATUS_SET,
+							data,
+							data_len,
+							error))
+			return FALSE;
+		if (!fu_nordic_hid_cfg_channel_cmd_receive(self, CONFIG_STATUS_SUCCESS, res, error))
+			return FALSE;
+
+		offset += data_len;
+	}
+
+	/* sync should return inactive for the last chunk */
+	if (is_last)
+		sync_state = DFU_STATE_INACTIVE;
+	return fu_nordic_hid_cfg_channel_dfu_sync(self, dfu_info, sync_state, error);
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_write_firmware_blob(FuNordicDeviceCfgChannel *self,
+					      GBytes *blob,
+					      FuProgress *progress,
+					      GError **error)
+{
+	g_autoptr(FuNordicCfgChannelDfuInfo) dfu_info = g_new0(FuNordicCfgChannelDfuInfo, 1);
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	if (!fu_nordic_hid_cfg_channel_dfu_sync(self, dfu_info, DFU_STATE_ACTIVE, error))
+		return FALSE;
+
+	chunks = fu_chunk_array_new_from_bytes(blob, 0, 0, dfu_info->sync_buffer_size);
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		gboolean is_last = (i == chunks->len - 1);
+		if (!fu_nordic_hid_cfg_channel_write_firmware_chunk(self, chk, is_last, error)) {
+			g_prefix_error(error, "chunk %u: ", fu_chunk_get_idx(chk));
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_write_firmware(FuDevice *device,
+					 FuFirmware *firmware,
+					 FuProgress *progress,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	FuNordicDeviceCfgChannel *self = FU_NORDIC_HID_CFG_CHANNEL(device);
+	guint32 checksum;
+	g_autofree gchar *image_id = NULL;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(FuNordicCfgChannelDfuInfo) dfu_info = g_new0(FuNordicCfgChannelDfuInfo, 1);
+
+	/* select correct firmware per target board, bootloader and bank */
+	image_id =
+	    g_strdup_printf("%s_%s_bank%01u", self->board_name, self->bl_name, self->flash_area_id);
+	firmware = fu_firmware_get_image_by_id(firmware, image_id, error);
+	if (firmware == NULL)
+		return FALSE;
+
+	checksum = g_ascii_strtoull(fu_firmware_get_checksum(firmware, 0, error), NULL, 16);
+	if (checksum == 0 && error != NULL)
+		return FALSE;
```

Comment:
This looks fishy. Maybe get the checksum as a `const gchar *csum` and then return FALSE if that fails. Then check for checksum=0 and do a `g_set_error()` in that case.

Reply:
ok, reworked a bit. However crc32 might be zero, afair, so have no sense to check it against 0.
Instead I set the '-1' for requested checksum kind, hence checksum will work only for the custom calculation function and added error handling to firmware's implementation.